lib/synapse/service_watcher/zookeeper/zookeeper.rb

```diff
@@ -113,38 +113,55 @@ def start_discovery
     end
 
     # find the current backends at the discovery path
-    def discover(zookeeper_opts={:watch => true})
+    def discover(zookeeper_opts={:watch => true}, zk_event=nil)
```

Comment:
@gmcatsf just to confirm, when reconnected but session timeout, it will pass `nil` event so that all watches will be set?

Reply:
Just to clarify, this method only handles registered node events and is not getting called when connection/session events occur.

In case of reconnected but session timeout, current event handling logic is to set @zk to nil, and then synapse will raise error and restart ZookeeperWatcher. At that time your statement is true, i.e. it passes nil event so all watches will be set.